src/typechecker.erl

```diff
@@ -20,6 +20,36 @@
 
 -type type() :: erl_parse:abstract_type().
 
+-type type_error() ::
+    #{error_type := {undef, call | record | user_type | remote_type | variable}
+                    | unexported_remote_type | var | char | atom
+                    | string | int | float | improper_tail | cons_pat
+                    | cons | nil | argument_length_mismatch | call_args
+                    | {call, args | arity | context} | mfa
+                    | return | expected_fun_type | op_type_too_precise
+                    | {arith_op, arguments | expression}
+                    | {int_op, arguments | expression}
+                    | non_number_argument_to_plus | non_number_argument_to_minus
+                    | non_boolean_argument_to_not
+                    | unary_op | non_integer_argument_to_bnot
+                    | logic_op | {rel_op, arguments | expression}
+                    | {list_op, arguments | expression}
+                    | operator_pattern | pattern | tuple
+                    | {bit, arguments | expression} | generator | bin_generator
+                    | check_clauses | record | record_pattern | record_update
+                    | {comprehension, bit_string | list | binary}
+                    | lambda | cyclic_type_vars,
+      line           => erl_anno:anno(),
+      module         => atom(),
+      name           => atom(),
+      arity          => non_neg_integer(),
+      actual_type    => type() | {type(), type()},
+      expected_type  => type(),
```

Comment:
The expected/actual type works well for type checking, when we have a known type which we want to check against an expression. But what about inference? Suppose we're checking the type of an addition expression and one of the operators turns out to be a tuple. What does these fields represent in this situation?

Reply:
1. It would be nice if fields with a name like ..._type is always a type(). Perhaps we can encode two types as a tuple type (`{type, _, tuple, [T1, T2]}`) to keep the structure, but a probably better suggestion is that we add another field `expected_operand_types :: [type()]` which can be used for binary operators.
2. I see that actual_type and expected_type are optional. That means the pretty-printing should be able to handle cases where these are missing, independently of type_error. This is probably fine. We can just make sure that pretty-printing is agnostic to this and just prints the fields that are present in the error map.